src/main/java/com/palantir/gradle/gitversion/GitVersionCacheService.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.gradle.gitversion;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.gradle.api.Project;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class GitVersionCacheService implements BuildService<BuildServiceParameters.None> {
+
+    private static final Logger log = LoggerFactory.getLogger(GitVersionCacheService.class);
+
+    private final Timer timer = new Timer();
+    private final ConcurrentMap<GitVersionArgs, VersionDetails> versionDetailsMap = new ConcurrentHashMap<>();
```

Comment:
I tried this out on a project and was wondering why it doesn't appear to produce a speedup on repos with large numbers of projects and calling `gitversion()` on each. I think this is the reason - `GitVersionArgs` does not implement `hashCode()` or `equals()` so a new entry is created on the map each time. I'd either just use `String` here (probably easiest), or make a new immutables type containing the string prefix and immutables will generate `hashCode()` and `equals()` for us.

Reply:
I switched back to using the concatenation of gitDir path and prefix as the key.

Reply:
If we do not use the gitDir as part of the key, we'd fail the `test subproject version` test where the two projects have different worktree.